### DISCUSSION TITLE
Function to change the specific music volume?

### DISCUSSION THREAD

**Discussion by chengenzhao** [AlmasB/FXGL#1200] [Ideas]
Hi:

We tried to find a way to tune the specific music volume
but seems like current API doesn't support this function
and we read the source code
seems like music.getAudio().setVolume() works
but this is internal API
so is it possible to provide a wrap function that allow developer to tune the specific music volume
or simply make music.getAudio() function open?

We currently have two background music
one is normal background music, the other one is white noise like train running noise or raining noise etc.
so some times we need to adjust just background music and keep the white noise running

Thanks
<img width="333" alt="截屏2022-08-19 上午11 30 40" src="https://user-images.githubusercontent.com/5525436/185536922-dbe436a7-4df6-4c52-a5bf-ecdee0885481.png">
<img width="663" alt="截屏2022-08-19 上午11 26 42" src="https://user-images.githubusercontent.com/5525436/185536590-11cd2baf-4967-4545-b036-1d3abef1bf68.png">

**Comment by chengenzhao** [comment]
I think a better solution is to mark each music/sound a label(id/key), maybe it could be a return string(id/key) of audioPlayer.loopMusic or .playMusic method
and then use this id/key to tune the volume
e.g.
```java
var id = FXGL.getAudioPlayer().loopMusic(bgm);
FXGL.getAudioPlayer().setVolume(id, 0.5);
```
and it would be nice to have a method to return the original volumeProperty 
```java
var volumeProperty = FXGL.getSettings().getMusicVolumeProperty(id); 
//or
var volumeProperty = FXGL.getSettings().getMusicVolumeProperty(music); 
```
then we could use Timeline to create a volume property transition
e.g.
```java
Timeline timeline = new Timeline(new KeyFrame(Duration.seconds(1), new KeyValue(getSettings().getMusicVolumeProperty(id), 0)));
timeline.play();
```

**Comment by AlmasB** [comment]
Thanks, I think this is covered in #1187 I'll mark it as a priority issue

**Reply by chengenzhao** [reply]
Good Evening

Happy to see the issue is labeled 17.3
and when the 17.2 will be released?
Since the new version of JavaFX 19.0 is about to release soon, in 3 weeks maybe
so 17.2 will support JFX 19 or still 18.0.2?

Thanks

**Reply by AlmasB** [reply]
Found time to push 17.2 (with jfx 18) out, so we can focus on 17.3

**Comment by DeathPhoenix22** [comment]
Hi @AlmasB & @chengenzhao,

I'm currently trying to integrate dynamic audio volumes for my game and since we don't have access to the volume property, I ended up doing this:

```
Music music = getAssetLoader().loadMusic("rain.mp3");
music.getAudio().setVolume(0.0);
DoubleProperty volumeHook = new SimpleDoubleProperty(0.0);
volumeHook.addListener((observableValue, number, t1) -> {
    music.getAudio().setVolume((double)t1);
});
Timeline timeline = new Timeline(new KeyFrame(Duration.seconds(5), new KeyValue(volumeHook, 1.0)));
FXGL.getAudioPlayer().playMusic(music);
timeline.play();
```

For now, this seems like an heavy workaround. It would be much easier to have access to the volume property. Also, I've looked at the code and having different sound volumes isn't possible as it will get reverted and replaced whenever a player do "Pause" because the unpause sequence resets volumes to master volumes.

I'll probably make a PR to resolve this :) Can I?

**Reply by DeathPhoenix22** [reply]
Hi, 

I've made a PR to allow volume control: #1311
Note: using music.getAudio() wouldn't be possible though as this was causing a bug with the settings.
